src/test/java/com/google/firebase/internal/RetryUnsuccessfulResponseHandlerTest.java

```diff
@@ -0,0 +1,281 @@
+
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+
+import com.google.api.client.http.HttpRequest;
+import com.google.api.client.http.HttpResponseException;
+import com.google.api.client.testing.http.FixedClock;
+import com.google.api.client.testing.util.MockSleeper;
+import com.google.api.client.util.Clock;
+import com.google.api.client.util.Sleeper;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.primitives.Longs;
+import com.google.firebase.testing.TestUtils;
+import java.io.IOException;
+import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.TimeZone;
+import org.junit.Test;
+
+public class RetryUnsuccessfulResponseHandlerTest {
+
+  private static final int MAX_RETRIES = 4;
+  private static final RetryConfig.Builder TEST_RETRY_CONFIG = RetryConfig.builder()
+      .setRetryStatusCodes(ImmutableList.of(429, 503))
+      .setMaxIntervalMillis(120 * 1000);
+
+  @Test
+  public void testDoesNotRetryOnUnspecifiedHttpStatus() throws IOException {
+    MultipleCallSleeper sleeper = new MultipleCallSleeper();
+    RetryUnsuccessfulResponseHandler handler = new RetryUnsuccessfulResponseHandler(
+        testRetryConfig(sleeper));
+    CountingLowLevelHttpRequest failingRequest = CountingLowLevelHttpRequest.fromStatus(404);
+    HttpRequest request = TestUtils.createRequest(failingRequest);
+    request.setUnsuccessfulResponseHandler(handler);
+    request.setNumberOfRetries(MAX_RETRIES);
+
+    try {
+      request.execute();
+      fail("No exception thrown for HTTP error");
+    } catch (HttpResponseException e) {
+      assertEquals(404, e.getStatusCode());
+    }
+
+    assertEquals(0, sleeper.getCount());
+    assertEquals(1, failingRequest.getCount());
+  }
+
+  @Test
+  public void testRetryOnHttpClientErrorWhenSpecified() throws IOException {
+    MultipleCallSleeper sleeper = new MultipleCallSleeper();
+    RetryUnsuccessfulResponseHandler handler = new RetryUnsuccessfulResponseHandler(
+        testRetryConfig(sleeper));
+    CountingLowLevelHttpRequest failingRequest = CountingLowLevelHttpRequest.fromStatus(429);
+    HttpRequest request = TestUtils.createRequest(failingRequest);
+    request.setUnsuccessfulResponseHandler(handler);
+    request.setNumberOfRetries(MAX_RETRIES);
+
+    try {
+      request.execute();
+      fail("No exception thrown for HTTP error");
+    } catch (HttpResponseException e) {
+      assertEquals(429, e.getStatusCode());
+    }
+
+    assertEquals(MAX_RETRIES, sleeper.getCount());
+    assertArrayEquals(new long[]{500, 1000, 2000, 4000}, sleeper.getDelays());
+    assertEquals(MAX_RETRIES + 1, failingRequest.getCount());
+  }
+
+
+  @Test
+  public void testRetryAfterIsAbsent() throws IOException {
+    MultipleCallSleeper sleeper = new MultipleCallSleeper();
```

Comment:
Cough cough :) This test is basically the same as the one above.

Reply:
Removed :)